src/System.Text.Json/src/System/Text/Json/JsonUtf8Reader.cs

```diff
@@ -0,0 +1,1274 @@
+        private bool HasMoreData()
+        {
+            if (_consumed >= (uint)_buffer.Length)
+            {
+                if (_isNotPrimitive && IsLastSpan)
+                {
+                    if (_tokenType != JsonTokenType.EndArray && _tokenType != JsonTokenType.EndObject)
+                        ThrowHelper.ThrowJsonReaderException(ref this, ExceptionResource.InvalidEndOfJson);
+                }
+                return false;
+            }
+            return true;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private bool HasMoreData(ExceptionResource resource)
+        {
+            if (_consumed >= (uint)_buffer.Length)
+            {
+                if (IsLastSpan)
+                    ThrowHelper.ThrowJsonReaderException(ref this, resource);
+                return false;
+            }
+            return true;
+        }
+
+        private bool ReadFirstToken(byte first)
+        {
+            if (first == JsonConstants.OpenBrace)
+            {
+                _currentDepth++;
+                _stackFreeContainer = 1;
+                _tokenType = JsonTokenType.StartObject;
+                _consumed++;
+                _lineBytePosition++;
+                _inObject = true;
+                _isNotPrimitive = true;
+            }
+            else if (first == JsonConstants.OpenBracket)
+            {
+                _currentDepth++;
+                _tokenType = JsonTokenType.StartArray;
+                _consumed++;
+                _lineBytePosition++;
+                _isNotPrimitive = true;
+            }
+            else
+            {
+                //Create local copy to avoid bounds checks.
+                ReadOnlySpan<byte> localCopy = _buffer;
+
+                if ((uint)(first - '0') <= '9' - '0' || first == '-')
+                {
+                    if (!TryGetNumber(localCopy.Slice(_consumed), out int numberOfBytes))
+                        return false;
+                    _tokenType = JsonTokenType.Number;
+                    _consumed += numberOfBytes;
+                    _lineBytePosition += numberOfBytes;
+                    goto Done;
+                }
+                else if (ConsumeValue(first))
+                    goto Done;
+
+                return false;
+
+            Done:
+                if (_consumed >= (uint)localCopy.Length)
+                    return true;
+
+                if (localCopy[_consumed] <= JsonConstants.Space)
+                {
+                    SkipWhiteSpace();
+                    if (_consumed >= (uint)localCopy.Length)
+                        return true;
+                }
+
+                if (_readerOptions.CommentHandling != JsonCommentHandling.Default)
+                {
+                    if (_readerOptions.CommentHandling == JsonCommentHandling.AllowComments)
+                    {
+                        if (_tokenType == JsonTokenType.Comment || localCopy[_consumed] == JsonConstants.Solidus)
```

Comment:
I'd like to see some comments describing why these two cases.  Presumably _tokenType was assigned by ConsumeValue, so describe what we already saw, and what are we looking at.